enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,389 @@
+
+### Goals
+
+* Operators can deliver console plugins separate from the console image and
+  update plugins when the operator updates.
+* The dynamic plugin API is similar to the static plugin API to ease migration.
+* Plugins can use shared console components such as list and details page components.
+* Shared components from core will be part of a well-defined plugin API.
+* Plugins can use [Patternfly 4](https://www.patternfly.org/v4/) components.
+* Cluster admins control what plugins are enabled.
+* Misbehaving plugins should not break console.
+* Existing static plugins are not affected and will continue to work as expected.
+* Plugins can declare the version of the console framework they support and
+  will be disabled when their version requirements are not met.
+
+### Non-Goals
+
+* Initially we don't plan to make this a public API. The target use is for Red
+  Hat operators. We might reevaluate later when dynamic plugins are more
+  mature.
+* We can't prevent breaking changes in Patternfly APIs the console exposes to plugins.
+* Plugins won't be sandboxed. They will have full JavaScript access to the DOM and network.
+* This proposal does not cover allowing plugins to contribute backend console endpoints.
+
+## Proposal
+
+### Module Federation
+
+Console will use [Webpack module federation](https://webpack.js.org/concepts/module-federation/)
+to load plugins at runtime. Module federation allows a JavaScript application
+to dynamically load code from another application while sharing dependencies.
+It also allows console to share its components with plugins built and bundled
+separately.
+
+Plugins will need to be built with Webpack 5 to use module federation.
+
+We have a pull request with a [working prototype](https://github.com/openshift/console/pull/6101)
+that uses module federation.
+
+### Shared Dependencies
+
+Console will expose React and Patternfly 4 as shared dependencies to plugins.
+Plugins can use any Patternfly component. Only a single version of React and
+Patternfly will be loaded.
+
+Plugins will not be able to use legacy Patternfly 3 components.
+
+### Delivering Plugins
+
+Plugins are delivered through operators. The operator will create a deployment
+on the platform with an HTTP server to host the plugin and expose it using a
+Kubernetes `Service`. The HTTP server serves all assets for the plugin,
+including JavaScript, CSS, and images. The `Service` must use HTTPS and a
+[service serving certificate](https://docs.openshift.com/container-platform/4.4/security/certificates/service-serving-certificate.html).
+The console backend will proxy the plugin assets from the `Service` using the
+service CA bundle.
+
+Operators declare that they have a console plugin available by creating a
+cluster-scoped `ConsolePlugin` resource that includes the service name, port,
+and path to the plugin manifest.
+
+```yaml
+apiVersion: console.openshift.io/v1alpha1
+kind: ConsolePlugin
+metadata:
+  name: my-plugin
+spec:
```

Comment:
I agree this would be a nicer API. My hesitation is that the ConsolePlugin resource is not owned by console. I could image the console operator and plugin operator fighting over updates to the resource.